### PR TITLE
feature/analytic module

### PR DESCRIPTION
**Pull request by xidealo**

### CODE REVIEW

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/confirm/ConfirmViewModel.kt [hunk 1]
```diff
@@ -104,6 +108,14 @@ class ConfirmViewModel(
     }
 
     private fun handleException(throwable: Throwable) {
+        analyticService.sendEvent(
+            ConfirmExceptionEvent, params = listOf(
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/confirm/ConfirmViewModel.kt:112]: add first param name and move second one to next line

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/consumercart/ConsumerCartViewModel.kt [hunk 2]
```diff
@@ -125,13 +133,39 @@ class ConsumerCartViewModel(
...
+        addProduct(
+            menuProductUuid = menuProductUuid
+        )
+    }
+
+    private fun addCartProductToCartClick(menuProductUuid: String) {
+        addProduct(
+            menuProductUuid = menuProductUuid
+        )
+        analyticService.sendEvent(CartAddEvent)
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/consumercart/ConsumerCartViewModel.kt:165]: - Let's rename this event to `AddCartProductToCartEvent`, this is really confusing now
- Why do we send `Add` event but don't send `Remove` event?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/product_details/ProductDetailsViewModel.kt [hunk 3]
```diff
@@ -47,14 +55,42 @@ class ProductDetailsViewModel(
...
                 addCartProductUseCase(menuProduct.uuid)
             }
         }
     }
 
+    private fun sendOnWantedClickedAnalytic(
+        menuProductUuid: String,
+        productDetailsOpenedFrom: ProductDetailsOpenedFrom,
+    ) {
+        when (productDetailsOpenedFrom) {
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/product_details/ProductDetailsViewModel.kt:76]: Let's make it shorter - add the `event` variable returned from `when` and then call `sendEvent` only once

File: analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.bunbeauty.analytic
+
+import com.bunbeauty.analytic.event.EventParameter
+import com.bunbeauty.analytic.event.FoodDeliveryEvent
+
+actual class AnalyticService {
+
+    actual fun sendEvent(event: FoodDeliveryEvent, params: List<EventParameter>){
+
```
Comment (MaxAstin) [inline analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt:9]: When are we going to implement it?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/product_details/ProductDetailsViewModel.kt [hunk 5]
```diff
@@ -47,14 +55,42 @@ class ProductDetailsViewModel(
...
+            ProductDetailsOpenedFrom.RECOMMENDATION_PRODUCT -> analyticService.sendEvent(
+                RecommendationAddEvent,
+                params = listOf(EventParameter("menuProductUuid", menuProductUuid))
+            )
+
+            ProductDetailsOpenedFrom.CART_PRODUCT -> analyticService.sendEvent(
+                CartAddEvent,
+                params = listOf(EventParameter("menuProductUuid", menuProductUuid))
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/product_details/ProductDetailsViewModel.kt:84]: - Let's add `EventParameter` child for all concrete param like `MenuProductUuidEventParameter` to be sure that all keys will be valid
- Let's add these params right to concrete events like `class CartAddEvent(menuProductUuidEventParameter: MenuProductUuidEventParameter)` to be sure that we will pass all necessary params

**Review by MaxAstin — CHANGES_REQUESTED** [review]

File: analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt (hunk 4 above)
Comment (xidealo) [reply analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt:9]: in feature/analytic_module_ios

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/confirm/ConfirmViewModel.kt [hunk 6]
```diff
@@ -104,6 +107,18 @@ class ConfirmViewModel(
...
 
     private fun handleException(throwable: Throwable) {
+        analyticService.sendEvent(
+            event = ConfirmErrorShowEvent(
+                error = when (throwable) {
+                    is TooManyRequestsException -> "TooManyRequests"
+                    is NoAttemptsException -> "NoAttempts"
+                    is InvalidCodeException -> "InvalidCode"
+                    is AuthSessionTimeoutException -> "AuthSessionTimeout"
+                    else -> "ShowSomethingWentWrong"
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/confirm/ConfirmViewModel.kt:117]: Without "Show"

**Review by MaxAstin — CHANGES_REQUESTED** [review]

File: analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.bunbeauty.analytic
+
+import cocoapods.FirebaseAnalytics.FIRAnalytics
+import com.bunbeauty.analytic.event.FoodDeliveryEvent
+import com.bunbeauty.analytic.event.toMap
+import platform.Foundation.NSBundle
+
+actual class AnalyticService {
+    actual fun sendEvent(event: FoodDeliveryEvent){
+        val targetName = NSBundle.mainBundle.objectForInfoDictionaryKey("CFBundleName")
```
Comment (MaxAstin) [inline analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt:10]: It's copy/paste. Let's move `targetName` from `:shared` module to `:core` module and connect to `:analytic` and `:shared` modules

File: app/src/main/java/com/bunbeauty/papakarlo/common/ui/animation/Animation.kt [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import androidx.compose.animation.core.tween
+import androidx.compose.animation.fadeIn
+import androidx.compose.animation.fadeOut
+import androidx.compose.animation.slideInVertically
+import androidx.compose.animation.slideOutVertically
+import androidx.compose.animation.with
+
+@OptIn(ExperimentalAnimationApi::class)
+val slideInAndSlideOutVerticallyWithFadeAnimation =
+    fadeIn(animationSpec = tween(220, delayMillis = 90)) + slideInVertically(
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/common/ui/animation/Animation.kt:13]: magical numbers, param names, copy/paste

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ui/CountPicker.kt [hunk 9]
```diff
@@ -47,12 +55,32 @@ fun CountPicker(
...
             onClick = onCountDecreased
         )
-        Text(
-            modifier = Modifier.padding(horizontal = 4.dp),
-            text = count.toString(),
-            style = FoodDeliveryTheme.typography.bodySmall.bold,
-            color = FoodDeliveryTheme.colors.mainColors.primary
-        )
+
+        AnimatedContent(
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ui/CountPicker.kt:59]: - Set `.height(IntrinsicSize.Min)` for container
- Remove `.height(36.dp)` for `AnimatedContent`
- Remove `contentAlignment = Alignment.Center` for `AnimatedContent`

**Review by MaxAstin — CHANGES_REQUESTED** [review]

File: analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt (hunk 7 above)
Comment (xidealo) [reply analytic/src/iosMain/kotlin/com/bunbeauty/analytic/AnalyticService.kt:10]: module for 1 varible?

**Review by MaxAstin — APPROVED** [review]

**Pull request merged**